# Algorithms behind Multi-Model Recommender

## Architecture of Prediction.io

![Alt](files/pio-architecture.png "Architecture")
<center><b><u>Prediction.io Architecture</u></b></center>

<p/>
## Demo

In [1]:
import numpy as np
import pandas
from IPython.display import display, HTML
from tabulate import tabulate

def displayTable(table, tableType, title = None):
    displayTables([table], [tableType], [title])
    
def displayTables(tables, tableTypes, titles = None, width = 60):
    if titles is None:
        titles = [None for i in range(len(tables))]
        
    html = '<table id="table123" border="0" cellpadding="10", width="' + \
                    str(width) + '%", align="center"><tr>'
    for table, tableType, title in zip(tables, tableTypes, titles):
        if tableType == "p": # Purchase history
            if title is None: title = "Purchase History:"
            row_labels = ['u'+str(i+1) for i in range(table.shape[0])]
            col_labels = ['t'+str(i+1) for i in range(table.shape[1])]

        elif tableType == 'c': # Co-occurrence matrix
            if title is None: title = "Co-occurrence:"
            row_labels = ['t'+str(i+1) for i in range(table.shape[1])]
            col_labels = row_labels

        elif tableType == 'cg': # Co-occurrence matrix with gender
            if title is None: title = "Co-occurrence with Gender:"
            row_labels = ['t'+str(i+1) for i in range(table.shape[0])]
            col_labels = ['M', 'F']

        elif tableType == 'v': # Just a vector
            if table.shape[0] > 1:
                table = table.transpose()
            row_labels = ['user']
            col_labels = ['t'+str(i+1) for i in range(table.shape[1])]
         
        elif tableType == 'g': # Gender info
            if title is None: title = "Gender:"
            row_labels = ['u'+str(i+1) for i in range(table.shape[0])]
            col_labels = ['M', 'F']

        html += "<td>"
        if title is not None:
            html += "<b>" + title + "</b><br/>"

        df = pandas.DataFrame(table, columns=col_labels, index=row_labels)
        html += df.to_html()
        
        html += "</td>"
    
    html += "</tr></table>"
    display(HTML(html))

## Purchase History Matrix and Co-Occurrence Matrix
Assume we have 5 users and 6 products.
- Matrix P will be the purchase history matrix
- P'P is the co-occurence matrix.
    - Each value represents the number of times two items co-occur

Examples:
- item 1 and item 3 co-occur twice (in user1 and user2)
- item 2 and item 4 co-occur twice (in user3 and user5)

In [2]:
P = np.matrix([[1,0,1,0,0,0], [1,0,1,0,1,0], [0,1,1,1,0,0], [0,1,0,1,0,1], [0,0,0,0,0,0]]) # History of purchase

PtP = P.transpose() * P
np.fill_diagonal(PtP, 0) # Fill the diagnonal with zeros, as they should be ignored.

displayTables([P, PtP], ['p', 'c'])


**Simple user-centric recommendation**

If a user is viewing item 1, what should we recommend?

In [3]:
h = np.matrix([1,0,0,0,0,0]).transpose()

r = PtP * h 
displayTable(r, 'v')
# It's the same as reading the 1st col of P'P

The above result tells us that we should recommend item 3, and then item 5.  It's based on the co-occurrence matrix:
- Two people who purchased item 1 also purchased item 3.
- One person who purchased item 1 also purchased item 5

This is similar to the "people who purchased this also purchase" list.

Now let's say a user comes in, and in her history she has purchased item 1 and item 2 in the past.  What should we recommend?

In [7]:
h = np.matrix([1,1,0,0,0,0]).transpose()
displayTables([P, PtP, PtP * h], ['p', 'c', 'v'], 
              [None, "P'P", "History: t1, t2<br/>Recs = P'P * h:"])
# It's the same as adding up the first two cols of P'P

## Cross Co-occurrence ##

In addition to purchase history, we also have Bookmark history of the 6 products.

**Cross-Cooccurrence Matrix**

In P'B, each **column** represent the bookmark's influence on purchase.  E.g. In the 4th column, you can see its influence on the purchases for t1, t3 and t5.

**Note**

These secondary event **doesn't** have to be any action which causes a purchase.  The event could be "dislike an item".  We don't need *causality* here.  The key thing is whether it 'cross co-occurs' with a primary event.


In [11]:
B = np.matrix([[0,0,0,1,0,0], [0,0,0,1,0,0], [0,1,0,0,0,0], [0,1,0,0,0,0], [1,0,0,0,1,1]]) # History of bookmark

PtB = P.transpose() * B

displayTables([P, B, PtB], ['p', 'p', 'c'], [None, 'Bookmark History:', "P'B:"], width = 75)

*Recommend a purchase based on bookmark history*

In [9]:
displayTables([P, B], ['p', 'p'], [None, 'Bookmark History:'], width=50)

print
print "Bookmark: 4   - P'B * b =", (PtB * np.matrix([0,0,0,1,0,0]).transpose()).transpose()
print "Bookmark: 1   - P'B * b =", (PtB * np.matrix([1,0,0,0,0,0]).transpose()).transpose()
print "Bookmark: 1&2 - P'B * b =", (PtB * np.matrix([1,1,0,0,0,0]).transpose()).transpose()


Bookmark: 4   - P'B * b = [[2 0 2 0 1 0]]
Bookmark: 1   - P'B * b = [[0 0 0 0 0 0]]
Bookmark: 1&2 - P'B * b = [[0 2 1 2 0 1]]


In the first case, u1 and u2 have bookmarked t4. So the recommendation is the same as adding the purchase history of u1 and u2.

In the last two cases, only user 5 have bookmarked items 1. But since he hasn't purchased any item, so we **cannot** correlate those this bookmark to any purchase. So item 1's bookmark has **no** effect on the recommendation. 

### Recommend a purchase based on both purchase history and bookmark history

In [10]:
displayTables([P, B], ['p', 'p'], [None, 'Bookmark History:'])

p = np.matrix([1,0,1,0,0,0]).transpose()
b = np.matrix([0,1,0,0,0,0]).transpose()
r = PtP*p + PtB*b

print
print "A user's purchase:", p.transpose()
print "Recommendations:  ", (PtP*p).transpose(), "\n"
print "A user's bookmark:", b.transpose()
print "Recommendations:  ", (PtB*b).transpose(), "\n"
print "Combined:         ", (PtP*p + PtB*b).transpose(), "\n"



A user's purchase: [[1 0 1 0 0 0]]
Recommendations:   [[2 1 2 1 2 0]] 

A user's bookmark: [[0 1 0 0 0 0]]
Recommendations:   [[0 2 1 2 0 1]] 

Combined:          [[2 3 3 3 2 1]] 



### What if we have users' gender information?

Let's say we have a new matrix, G, which record the gender information.  The columns represent Male and Female.

P'G is the cross-cooccurrence matrix.  Each column represents the gender's influence on purchase.

In [24]:
G = np.matrix([[1,0], [0,1], [0,1], [1,0], [0,1]])
PtG = P.transpose() * G
displayTables([P, G, PtG], ['p', 'g', 'cg'], [None, None, "P'G"], width=40)


*For a female customer, what should we recommend?*

In [60]:
g = np.matrix([0,1]).transpose()
print "For a female, recommendation = P'G*g =", (PtG * g).transpose()

For a female, recommendation = P'G*g = [[1 1 2 1 1 0]]


### Recommend based on purchase history, bookmark history and gender info

In [25]:
displayTables([P, B, PtG], ['p', 'p', 'cg'], [None, "Bookmark History:", "P'G"], width=75)

print "A user's purchase:", p.transpose()
print "Recommendations:  ", (PtP*p).transpose(), "\n"
print "A user's bookmark:", b.transpose()
print "Recommendations:  ", (PtB*b).transpose(), "\n"
print "A user's gender:  ", g.transpose()
print "Recommendations:  ", (PtG*g).transpose(), "\n"

print "Combined (all):   ", (PtP*p + PtB*b + PtG*g).transpose(), "\n"


A user's purchase: [[1 0 1 0 0 0]]
Recommendations:   [[2 1 2 1 2 0]] 

A user's bookmark: [[0 1 0 0 0 0]]
Recommendations:   [[0 2 1 2 0 1]] 

A user's gender:  

NameError: name 'g' is not defined

## Similar Items

**Warning**:
This section is not complete and is WIP. For item 4, from P'P we see that it co-occurs with item 2, 3 and 6 (in descending strength order).

I originally thought "hey, they occur with item 4, so that will be the similar items".

But Pat Ferrel pointed out that in fact item 4's "similar items" should be the rows which are similar to the 4th row in P'P.  I think this makes sense intuitively because items similar to item 4 should share similar co-occurence behavior with it.  Just co-occurs with it (like my original thought) is doing only a market basket analysis.

In [43]:
displayTables([P,PtP], ['p', 'c'])

** Under construction **

In [12]:
#t = np.matrix([0,0,0,1,0,0]).transpose()
#print (PtP*t).transpose()

## Summary \#1
- Purchase is our main event.
- Matrix **P** contains the history of all user purchases.
- We have have secondary "events": **B** - Bookmark, and **G** - Gender 
- Let **p**, **b** and **g** be *vectors* representing the purchase history, current bookmarks and gender of a user
- Recommendation:  `r = P'P p + P'B b + P'G g + ...`

### Sidetrack 1: Log-Likelihood Ratio (LLR) Analysis

A key characteristic of a successful recommender engine is serendipty.

**Serendipity** - Serendipity is a measure "how surprising the recommendations are". For instance, a recommender system that recommends milk to a customer in a grocery store, might be perfectly accurate but still it is not a good recommendation because it is an obvious item for the customer to buy.

In our co-occurrence matrix, we want to filter out pairs which are *not* surprising.


In [13]:
def printSubTotalTable(table, headers, rows):
    table[2,0] = np.sum(table[:,0]) 
    table[2,1] = np.sum(table[:,1]) 
    table[0,2] = np.sum(table[0,:])
    table[1,2] = np.sum(table[1,:])
    
    tableNew = np.concatenate((np.array(rows).reshape(-1,1), table), axis=1)
    tableNew[2,3] = ''
    print tabulate(tableNew, headers, tablefmt='grid')
    print "%% of cake-eater who drinks whisky = %d/%d = %.2f%%" % \
        (table[0,0], table[2,0], (table[0,0]*1.0/table[2,0]*100))
    print "%% of population who drinks whisky = %d/%d = %.2f%%" % \
        (table[0,2], (table[0,2]+table[1,2]), (table[0,2]*1.0/(table[0,2]+table[1,2])*100))
    
headers = ['', 'Cake', 'Others', '']
rows = ['Whisky', 'Others', '']

case1 = np.array(
    [
        [13, 1000, 0],
        [1000, 100000, 0],
        [0, 0, 0]
    ])

print "Case 1"
printSubTotalTable(case1, headers, rows)

print
print "Case 2"
case2 = np.array(
    [
        [10, 3, 0],
        [2, 100000, 0],
        [0, 0, 0]
    ])
printSubTotalTable(case2, headers, rows)


Case 1
+--------+--------+----------+--------+
|        |   Cake |   Others |        |
+========+========+==========+========+
| Whisky |     13 |     1000 | 1013   |
+--------+--------+----------+--------+
| Others |   1000 |   100000 | 101000 |
+--------+--------+----------+--------+
|        |   1013 |   101000 |        |
+--------+--------+----------+--------+
% of cake-eater who drinks whisky = 13/1013 = 1.28%
% of population who drinks whisky = 1013/102013 = 0.99%

Case 2
+--------+--------+----------+--------+
|        |   Cake |   Others |        |
+========+========+==========+========+
| Whisky |     10 |        3 | 13     |
+--------+--------+----------+--------+
| Others |      2 |   100000 | 100002 |
+--------+--------+----------+--------+
|        |     12 |   100003 |        |
+--------+--------+----------+--------+
% of cake-eater who drinks whisky = 10/12 = 83.33%
% of population who drinks whisky = 13/100015 = 0.01%


** When will information become interesting? **

In Case 1, among cake-lovers, the fraction of whisky-drinker is similar to that fraction in the whole population.  So 1.28% doesn't provide any new info.

In Case 2, the situation is very different.  In the population, only 0.01% drinks whisky, but among cake-lovers the fraction is 83%.  So we found something very interesting.

The engine uses a test called *Log-Likelihood Ratio Analysis (LLR)* to **filter** out all non-interesting co-occurrence:<br/>
- All non-interesting cases become zero
- All interesting cases become one

Details: http://tdunning.blogspot.hk/2008/03/surprise-and-coincidence.html

So our co-occurrence matrix P'P may become like this:

In [45]:
PtP_llr = np.array([[0,0,0,0,0,1],[0,0,1,1,0,0],[0,1,0,0,0,0],[0,1,0,0,0,1],[1,0,0,0,0,0],[1,0,0,1,0,0]])
displayTables([PtP, PtP_llr], ['c','c'], ["P'P", "LLR(P'P)"])

**The recommendation equation becomes:**<br/>
`r = LLR(P'P) p + LLR(P'B) b + LLR(P'G) g`

### Sidetrack 2: kNN using [cosine distance][1] and Search Engine
[1]: http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.spatial.distance.cosine.html "Cosine Distance"

In [22]:
# Cosine distance notes:
from scipy.spatial import distance

v_target = np.array([0,1,1,0])

print "%.3f" % distance.cosine(v_target, [1,1,1,0])
print "%.3f" % distance.cosine(v_target, [1,1,0,0])
print "%.3f" % distance.cosine(v_target, [1,1,1,1])
print "%.3f" % distance.cosine(v_target, [0,0,1,0])


0.184
0.500
0.293
0.293


In [16]:
h = np.matrix([0,0,1,0,1,0]).transpose()
displayTables([PtP, h, PtP*h], ['c', 'v', 'v'], \
              ["P'P", "A user's purchase history", "Recs = P'P * h"], width=90)

print
print "h =", h.transpose()
from scipy.spatial import distance
for r in range(6):
    dist = distance.cosine(h, PtP[r])
    print "Item", r + 1, "- cosine distance with h =", ("-" if h[r,0] == 1 else dist)


h = [[0 0 1 0 1 0]]
Item 1 - cosine distance with h = 0.0513167019495
Item 2 - cosine distance with h = 0.711324865405
Item 3 - cosine distance with h = -
Item 4 - cosine distance with h = 0.711324865405
Item 5 - cosine distance with h = -
Item 6 - cosine distance with h = 1.0


Based on nearest neighbour, the three closest items are 1, 2, and 4.  That matches our earlier results.

**Intuition:** <br/>
- To get P'P * h, we add up all the columns of t3 and t5 in P'P, and the recs are those with large values, which means they co-occurs many times with t3 and also t5.
  - In descending order they are t1, t2 and t4. (t1 is the top)
- In the LLR(P'P) case, LLR(P'P) contains only 0 and 1, with 1 meaning two items have good-quality co-occurrence.  
- If t1 has such quality with t3 and t5:
  - The ideal vector in LLR(P'P) should be `[0,0,1,0,1,0]`, which has zero distance.
  - A less ideal case could be `[1,0,1,0,1,0]`, which is still close, but with some noise.


## Search Engine as prediction engine
Remember our recommendation formula:
`r = LLR(P'P) p + LLR(P'B) b + LLR(P'G) g + ...`

The Universal Recommender actually stores `LLR(P'P), LLR(P'B) and LLR(P'G)` as three searcheable fields in Elastic Search:
- Each item is a doc in ES.
- For each item, UR stores its corresponding row from the indicator matrices (e.g. LLR(P'G)) in a field in the ES index.
- Item's properties (e.g. categories, price) are also stored as fields of the doc.
- Using search engine we have lots of flexibilities in boosting and filtering.

** Demo **

## Summary \#2
- Purchase is our main event.
- Matrix P contains the history of all user purchases.
- We have have secondary "events": B - Bookmark, and G - Gender, **and more....**
- Let a, b and g be *vectors* representing the purchase history, current bookmarks and gender of a user
- Recommendation:  `r = LLR(P'P) p + LLR(P'B) b + LLR(P'G) g + ...`
- We can define properties for our item:
    - Price
    - Brand
    - etc.
- Using Search Engine, we can perform filtering and boosting.  E.g.:
    - Boost a certain brand
    - Boost a certain price range
    - Limit to a certain price range
    
## Evaluation
- Pros
    - Open Source - can study and modify the source code.
    - The algorithm was suggested by Ted Dunning
        - Chief Application Architect at MapR
        - Ted was the chief architect behind the MusicMatch (now Yahoo Music) and Veoh recommendation systems.
- Cons
    - The Universal Recommender is relatively new - v0.2 released in July, now at v0.2.2
    - Primary event: need to choose a good subset of events from Price.com
    - Not enough data for testing.  Need to wait for more.